sympy/series/formal.py

```diff
@@ -0,0 +1,445 @@
+        if self.ind:
+            for t in Add.make_args(self.ind):
+                xterm = _get_xterm(t)
+                if pt_xk == 1 and xterm == 1:
+                    ind += t
+                elif xterm != 1 and pt_xk != 1:
+                    j = xterm.as_base_exp()[1]
+                    if j == pt:
+                        ind += t
+
+        try:
+            pt_ak = self.ak.coeff(pt).simplify() # TODO: Thoughts?
+        except IndexError:
+            pt_ak = S.Zero
+
+        term = (pt_ak * pt_xk) + ind
+
+        return term.collect(self.x)
+
+    def _eval_subs(self, old, new):
+        x = self.x
+        if old.has(x):
+            return self
+
+    def _eval_as_leading_term(self, x):
+        for t in self:
+            if t is not S.Zero:
+                return t
+
+
+def fps(f, x=None, x0=0, dir=1, hyper=True, order=4, rational=True, full=False):
+    """Generates Formal Power Series of f
+
+    returns the formal series expansion of ``f`` around ``x = x0``
+    with respect to ``x`` in the form of a ``FormalPowerSeries`` object
+
+    Formal Power Series is represented using an explicit formula
+    computed using different algorithms.
+
+    See :func:`compute_fps` for the more details regarding the computation
+    of formula.
+
+    Arguments
+    =========
+
+    * x = if ``x=None`` and ``f`` is univariate, the univariate
+    symbols will be supplied, otherwise an error will be raised.
+
+    * x0 = series expansion around ``x = x0``(Default = 0).
+
+    * dir = For ``dir=1`` (default) the series is calculated from the right and
+    for ``dir=-1`` the series from the left. For smooth functions this
+    flag will not alter the results.
+
+    * hyper = set hyper=False, for not using hypergeometric algorithm.
+
+    * order = Order of the derivative of f, till which algorithms
+    are run.
+
+    * rational = set rational=False to skip rational algorithm
+
+    * full = full by default is False. Try setting full to True
+    to increase the range of rational algorithm. See :func:`rational_algorithm`
+    for details.
+
+    Examples
+    ========
+
+    >>> from sympy import fps, O, ln, atan
+    >>> from sympy.abc import x
+
+    Rational Functions
+
+    >>> fps(ln(1 + x)).truncate()
+    x - x**2/2 + x**3/3 - x**4/4 + x**5/5 + O(x**6)
+
+    >>> fps(atan(x), full=True).truncate()
+    x - x**3/3 + x**5/5 + O(x**6)
+
+    See Also
+    ========
+
+    sympy.series.formal.FormalPowerSeries
+    sympy.series.formal.compute_fps
+    """
+    f= sympify(f)
+
+    if x is None:
+        free = f.free_symbols
+        if len(free) == 1:
+            x = free.pop()
+        elif not free:
+            return f
+        else:
+            raise NotImplementedError("multivariate formal power series")
+    else:
+        x = sympify(x)
+
+    if not f.has(x):
+        return f
+
+    x0 = sympify(x0)
+    dir = sympify(dir)
+
+    if dir not in [S.One, -S.One]:
+        raise ValueError("Dir must be 1 or -1")
```

Comment:
Most of these sympifications/input checks should probably be moved to `compute_fps`. The recursive nature could be covered by renaming the current function to `_compute_fps`, and creating a light wrapper that validates/sympifys inputs, then calls the actual computing method.